src/backend/distributed/connection/connection_management.c

```diff
@@ -310,21 +312,68 @@ StartNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port,
 	if (!(flags & FORCE_NEW_CONNECTION))
 	{
 		/* check connection cache for a connection that's not already in use */
-		connection = FindAvailableConnection(entry->connections, flags);
+		MultiConnection *connection = FindAvailableConnection(entry->connections, flags);
 		if (connection)
 		{
 			return connection;
 		}
 	}
 
+
 	/*
 	 * Either no caching desired, or no pre-established, non-claimed,
 	 * connection present. Initiate connection establishment.
 	 */
+	MultiConnection *connection = MemoryContextAllocZero(ConnectionContext,
+														 sizeof(MultiConnection));
```

Comment:
We might see this type of dummy connection in AfterXactHostConnectionHandling. That looks fine, but I'm wondering whether it should be made explicit with a flag.

Reply:
@marcocitus  Please see 40f5a0a9b74fa610df152c73204d2b46486b1e8e, I'm not merging the PR yet, waiting your comment on that.

Basically, expand `sharedCounterIncremented ` into a state, and handle it `AfterXactHostConnectionHandling ` as well.

>  That looks fine, 

I also think that the behavior is the same, so please make the call on which you prefer. I've a slight preference to have the `state`, but didn't like the naming very much -- but couldn't come up something better. So, any suggestion on that also could be helpful.

Reply:
I like that commit. The enum names should perhaps have a common prefix (POOL_STATE_NOT_INITIALIZED, POOL_STATE_COUNTER_INCREMENTED, POOL_STATE_INITIALIZED).